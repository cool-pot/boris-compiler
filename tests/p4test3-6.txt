*** test level 3, function: tuple->int
global t = 1,2,3;
defun sum(x)
    return x.0 + x.1 +x.2;
end defun

print sum t;
print sum (3,4,5,6,7);***only sum first three