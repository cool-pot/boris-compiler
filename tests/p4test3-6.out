; ModuleID = 'module-boris-input'
source_filename = "module-boris-input"
target triple = "x86_64-pc-linux-gnu"

@format_int = private unnamed_addr constant [4 x i8] c"%d\0A\00"
@format_int.1 = private unnamed_addr constant [4 x i8] c"%d\0A\00"

define i32 @strlen(i8* %s) {
init:
  %i = alloca i32
  store i32 0, i32* %i
  br label %check

check:                                            ; preds = %body, %init
  %0 = load i32, i32* %i
  %1 = getelementptr i8, i8* %s, i32 %0
  %2 = load i8, i8* %1
  %3 = icmp ne i8 0, %2
  br i1 %3, label %body, label %end

body:                                             ; preds = %check
  %4 = load i32, i32* %i
  %5 = add i32 %4, 1
  store i32 %5, i32* %i
  br label %check

end:                                              ; preds = %check
  %6 = load i32, i32* %i
  ret i32 %6
}

declare i32 @printf(...)

define void @main() {
MainEntry:
  %t = alloca i32, i32 3
  %0 = getelementptr i32, i32* %t, i32 0
  store i32 1, i32* %0
  %1 = getelementptr i32, i32* %t, i32 1
  store i32 2, i32* %1
  %2 = getelementptr i32, i32* %t, i32 2
  store i32 3, i32* %2
  %3 = call i32 @sum(i32* %t)
  %4 = call i32 (...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @format_int, i32 0, i32 0), i32 %3)
  %5 = alloca i32, i32 5
  %6 = getelementptr i32, i32* %5, i32 0
  store i32 3, i32* %6
  %7 = getelementptr i32, i32* %5, i32 1
  store i32 4, i32* %7
  %8 = getelementptr i32, i32* %5, i32 2
  store i32 5, i32* %8
  %9 = getelementptr i32, i32* %5, i32 3
  store i32 6, i32* %9
  %10 = getelementptr i32, i32* %5, i32 4
  store i32 7, i32* %10
  %11 = call i32 @sum(i32* %5)
  %12 = call i32 (...) @printf(i8* getelementptr inbounds ([4 x i8], [4 x i8]* @format_int.1, i32 0, i32 0), i32 %11)
  br label %EndEntry

EndEntry:                                         ; preds = %MainEntry
  ret void
}

define i32 @sum(i32*) {
Entry:
  %1 = getelementptr i32, i32* %0, i32 0
  %load_tuple_ref_int_temp = load i32, i32* %1
  %2 = getelementptr i32, i32* %0, i32 1
  %load_tuple_ref_int_temp1 = load i32, i32* %2
  %expr_plus_expr_tmp = add i32 %load_tuple_ref_int_temp, %load_tuple_ref_int_temp1
  %3 = getelementptr i32, i32* %0, i32 2
  %load_tuple_ref_int_temp2 = load i32, i32* %3
  %expr_plus_expr_tmp3 = add i32 %expr_plus_expr_tmp, %load_tuple_ref_int_temp2
  ret i32 %expr_plus_expr_tmp3
}
